renter/programbuilder.go

```diff
@@ -0,0 +1,269 @@
+package renter
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+
+	"go.sia.tech/core/net/rhp"
+	"go.sia.tech/core/types"
+)
+
+// A ProgramBuilder constructs programs for the renter to execute on a host.
+type ProgramBuilder struct {
+	instructions []rhp.Instruction
+	readwrite    bool
+	data         *bytes.Buffer
+	usage        rhp.ResourceUsage
+
+	duration uint64
+	settings types.HostSettings
+
+	err error // sticky error
+}
+
+func (pb *ProgramBuilder) addUsage(usage rhp.ResourceUsage) {
+	pb.usage.Add(usage)
+}
+
+func (pb *ProgramBuilder) appendInstruction(id [16]byte, args []byte) {
+	switch id {
+	case rhp.InstrAppendSector, rhp.InstrDropSectors, rhp.InstrSwapSector:
+		pb.readwrite = true
+	}
+
+	pb.instructions = append(pb.instructions, rhp.Instruction{
+		Specifier: id,
+		Arguments: args,
+	})
+}
+
+func (pb *ProgramBuilder) setErr(err error) error {
+	if pb.err == nil {
+		pb.err = err
+	}
+	return pb.err
+}
+
+// AddAppendSectorInstruction adds an append sector instruction to the program.
+func (pb *ProgramBuilder) AddAppendSectorInstruction(sector *[rhp.SectorSize]byte, proof bool) *ProgramBuilder {
```

Comment:
"builder" objects returning themselves so that you can chain methods together is a JavaScript-ism. No real need for it here.